db/db_impl/db_impl.cc

```diff
@@ -1445,6 +1447,15 @@ Status DBImpl::ApplyReplicationLogRecord(
           cfd->InstallSuperVersion(&sv_context, &mutex_);
         }
 
+        if (evictObsoleteFiles) {
+          versions_->GetObsoleteFiles(&job_context.sst_delete_files,
+                                      &job_context.blob_delete_files,
+                                      &job_context.manifest_delete_files,
+                                      std::numeric_limits<uint64_t>::max());
+          versions_->RemoveLiveFiles(job_context.sst_delete_files,
```

Comment:
I don't quite understand how this function works. Based on the implementation, it checks existence of file number in `file_locations_` hashmap to decide whether we should close the file or not. But based on code search, seems we never erase from the `file_locations_` hashmap?

```
> egrep file_locations_ . -nir
./version_set.cc:3957:  file_locations_.reserve(num_files);
./version_set.cc:3966:      assert(file_locations_.find(file_number) == file_locations_.end());
./version_set.cc:3967:      file_locations_.emplace(file_number, FileLocation(level, pos));
./version_set.cc:5963:      vstorage->file_locations_[file_number] =
./version_set.h:350:    const auto it = file_locations_.find(file_number);
./version_set.h:352:    if (it == file_locations_.end()) {
./version_set.h:623:  FileLocations file_locations_;
```

Reply:
`file_locations_` is per version, is regenerated for each version in `VersionStorageInfo::GenerateFileLocationIndex`, it's basically a secondary index for `files_`. `RemoveLiveFiles` checks if the file is still alive in the latest version, which could be the case with trivial move -- we have two file metadatas pointing to the same file, in which case we should delete the file metadata, but not delete the file or evict it from the table cache.

Reply:
Makes sense!